### PR TITLE
feat(pgvector): mmr_search — diversity-aware re-ranking (MMR)

### PR DESCRIPTION
**Pull request by devopam**
## Why

**Item 4 of the feature plan, first slice** (shortlist **9.2**), following the agreed one-tool-per-PR cadence for the pgvector expansion.

## What

**`mmr_search`** — diversity-aware vector search. Plain k-NN tends to return near-duplicates; MMR trades a little relevance for variety, which usually improves the context an agent feeds to an LLM.

The algorithm fetches `fetch_k` nearest candidates by pgvector distance (the recall pass), then greedily picks `k` maximising:

```
lambda·sim(query, doc) - (1-lambda)·max sim(doc, already_picked)
```

- `lambda_mult=1` → pure relevance (≈ `vector_search`); `0` → pure diversity; default `0.5`.
- Relevance **and** the diversity penalty are both cosine similarities computed **in-process over the candidate embeddings**, so the ranking is independent of which `metric` drove the recall pass.
- `fetch_k` defaults to `max(4·k, 20)`.
- Each hit carries `relevance`, `mmr_score`, and selection `rank`. Read-only; `available=false` without pgvector.

Implementation: new `MmrMatch` / `MmrSearchResult` dataclasses + `mmr_search()` in `mcpg.textsearch`, two small pure helpers (`_parse_embedding` — accepts pgvector's list form *or* its `"[..]"` text literal; `_cosine_similarity` — zero-vector safe), and the tool registration in `tools.py`.

## Test plan
- [x] **+15 tests** in `test_textsearch.py`: unavailable-without-pgvector; **pure-relevance keeps the near-duplicate**; **pure-diversity swaps in the distinct row** (the core MMR behaviour, asserted on a hand-built query/A/B/C set); embedding column dropped from result rows; bracketed-text embedding parsing; default `fetch_k` pool sizing; argument validation (`metric` / `k` / `lambda_mult` / `fetch_k>=k` / non-finite); tool callable from a client; plus direct coverage of both helpers.
- [x] **1024 unit tests pass**; ruff / format / mypy(`src/mcpg`) / bandit clean.

## Docs (folded in)
- `tour.md` Search section; `feature-shortlist.md` 9.2 → ✅; CHANGELOG.
- Tool count **121 → 122** across `tour` / `user-guide` / `tools`.

## Remaining pgvector slices
`import_vectors` (9.6, next), then `detect_vector_outliers` (9.4) + `cluster_vectors` (9.3), as separate focused PRs.

---
_Generated by [Claude Code](https://claude.ai/code/session_0122yLZLJ8t4W43sdN6BmTZc)_

## Summary by Sourcery

Add a diversity-aware pgvector MMR search tool and expose it via the MCPg tool interface.

New Features:
- Introduce MmrMatch and MmrSearchResult result types for Maximal Marginal Relevance vector searches.
- Add an mmr_search function that re-ranks pgvector k-NN results for diversity using a lambda-controlled relevance/diversity trade-off and returns per-hit relevance, MMR score, and rank.
- Expose mmr_search as a new MCPg server tool, returning structured results and reporting availability based on the pgvector extension.

Documentation:
- Document the new mmr_search tool in the changelog, tool tour, and tools reference, and update the advertised tool count to 122.
- Mark the mmr_search feature as shipped in the feature shortlist, noting its relevance/diversity trade-off and metric-independent behaviour.

Tests:
- Add unit tests covering MMR behaviour across relevance-only and diversity-only settings, result shaping, default fetch_k sizing, argument validation, and non-finite query vectors.
- Add tests for the embedding parsing and cosine similarity helpers, including multiple input formats and zero-vector handling.
- Add an integration-style test ensuring the mmr_search tool is discoverable and callable from a client session.

### CODE REVIEW

File: src/mcpg/textsearch.py [hunk 1]
```diff
@@ -311,6 +339,137 @@ async def vector_range_search(
...
+    )
+
+    candidates: list[tuple[dict[str, Any], list[float], float]] = []
+    for row in rows or []:
+        cells = dict(row.cells)
+        cells.pop("mcpg_distance", None)
+        embedding = _parse_embedding(cells.get(column))
+        cells.pop(column, None)  # drop the embedding from the returned row
+        relevance = _cosine_similarity(query_vector, embedding)
+        candidates.append((cells, embedding, relevance))
```
[the remark this answers is not in the document]
Comment (devopam) [reply src/mcpg/textsearch.py:468]: Good catch — fixed in b7543e8. A row whose embedding cell is `NULL` is now skipped before parsing, so a partially-embedded table doesn't fail the whole search.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_textsearch.py [hunk 2]
```diff
@@ -226,6 +228,149 @@ async def test_vector_search_tool_is_callable_from_a_client() -> None:
...
+    routes = {
+        "pg_extension": [{"present": 1}],
+        "::vector": [
+            {"id": 1, "embedding": "[1.0,0.0]", "mcpg_distance": 0.0},
+            {"id": 2, "embedding": "[0.0,1.0]", "mcpg_distance": 1.0},
+        ],
+    }
+    result = await mmr_search(FakeRoutingDriver(routes), "app", "docs", "embedding", [1.0, 0.0], k=2)  # type: ignore[arg-type]
+
+    assert [m.row["id"] for m in result.matches] == [1, 2]
```
[the remark this answers is not in the document]
Comment (devopam) [reply tests/unit/test_textsearch.py:296]: Added in b7543e8 — `test_mmr_search_skips_rows_with_null_embeddings` routes a `None` embedding through the candidate set and asserts the row is dropped while the rest are reranked normally.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/mcpg/textsearch.py [hunk 3]
```diff
@@ -311,6 +339,137 @@ async def vector_range_search(
...
+        return [float(part) for part in inner.split(",")]
+    raise SearchError(f"could not parse embedding value of type {type(value).__name__}")
+
+
+def _cosine_similarity(a: list[float], b: list[float]) -> float:
+    """Cosine similarity of two equal-length vectors; 0.0 if either is zero."""
+    dot = 0.0
+    norm_a = 0.0
+    norm_b = 0.0
+    for x, y in zip(a, b, strict=False):
```
[the remark this answers is not in the document]
Comment (devopam) [reply src/mcpg/textsearch.py]: Done in b7543e8 — `_cosine_similarity` now uses `zip(..., strict=True)`, and `mmr_search` does a friendly upstream check on each candidate's parsed embedding (`SearchError(f"row's {column!r} embedding has dimension N, expected M ...")`), so callers get a clear, row-context error instead of an obscure `ValueError` from `zip`. New tests cover both layers: a 3-dim row vs 2-dim query raises the friendly error, and the direct `_cosine_similarity([1, 0], [1, 0, 0])` raises (as the low-level invariant).

Skipped overall #2 (O(k²) `remaining.remove()`): the MMR loop is dominated by the cosine work inside it — O(k · fetch_k · d) with d typically 768–4096. A bitmask doesn't move the needle and would add complexity. Happy to revisit if a profile ever flags it.

---
_Generated by [Claude Code](https://claude.ai/code)_
